### PR TITLE
Converts first LibraryAnalysis repository to a benchbuild project

### PR DESCRIPTION
**Pull request by cormensratio**
Converts the TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject repository to a benchbuild project

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/elementalist.py  3
         

Clones added
============
- varats/projects/cpp_projects/elementalist.py  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by cormensratio** [comment]
Codacy: Should I try to find another name for the whole project, since the name does not conform with snake style naming?

- Can I assume that the current project state works if just_compile finishes without an error?

**Comment by vulder** [comment]
> Codacy: Should I try to find another name for the whole project, since the name does not conform with snake style naming?
> 
> * Can I assume that the current project state works if just_compile finishes without an error?

Yes, for the class and module name you should choose a valid name.

If just_compile finishes without errors it normally works. Did you get a working binary?
You can test this by setting `clean` in the benchbuild config to false. Afterwards, go to the `benchbuild/results/just_compile/projectname@ref/` dir and check if the binary works.

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+class TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject(
+    bb.Project
+):  # type: ignore
+    """Class to analyse interactions between two discrete libraries and one
+    project."""
+
+    NAME = 'TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject'
+    GROUP = 'cpp_projects'
+    DOMAIN = 'analysis'
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]:
```suggestion
    DOMAIN = 'library-testproject'
```

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            version_filter=project_filter_generator(
+                "TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject"
+            )
+        )
+    ]
+
+    @property
+    def binaries(self) -> tp.List[ProjectBinaryWrapper]:
+        """Return a list of binaries generated by the project."""
+        return wrap_paths_to_binaries(["elementalist"])
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]: Is this really the full path to the generated binary?

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    def run_tests(self) -> None:
+        pass
+
+    def compile(self) -> None:
+        """Contains instructions on how to build the project."""
+
+        git_path = get_local_project_git_path(self.NAME)
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]:
```suggestion
        library_source_folder = get_local_project_git_path(self.NAME)
```

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            with local.env(CC=str(clang)):
+                bb.watch(cmake)("-G", "Unix Makefiles", ".")
+            bb.watch(make)("-j", get_number_of_jobs(bb_cfg()))
+
+    def __compile_make(self) -> None:
+        version_source = bb.path(self.source_of_primary)
+        clang = bb.compiler.cc(self)
+        with local.cwd(version_source):
+            with local.env(CC=str(clang)):
+                bb.watch(make)("-j", get_number_of_jobs(bb_cfg()))
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]: Why do you need a special compile function for make? If I remember correctly, you only support cmake in your lib.

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    def __compile_cmake(self) -> None:
+        version_source = bb.path(self.source_of_primary)
+        clang = bb.compiler.cc(self)
+        with local.cwd(version_source):
+            with local.env(CC=str(clang)):
+                bb.watch(cmake)("-G", "Unix Makefiles", ".")
+            bb.watch(make)("-j", get_number_of_jobs(bb_cfg()))
+
+    def __compile_make(self) -> None:
+        version_source = bb.path(self.source_of_primary)
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]:
```suggestion
        version_source = local.path(self.source_of_primary)
```

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                "cc6a93677f9a9825d3bbe20f1201e6652ba4a7d", "master"
+            )
+
+        if version in cmake_revisions:
+            self.__compile_cmake()
+        else:
+            self.__compile_make()
+
+    def __compile_cmake(self) -> None:
+        version_source = bb.path(self.source_of_primary)
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]:
```suggestion
        version_source = local.path(self.source_of_primary)
```

**Review by vulder — CHANGES_REQUESTED** [review]
Found a fews things, did you just copy the compile function? This seems a bit off for your library.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         

Clones added
============
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Comment by cormensratio** [comment]
As long as multiple VaraTestRepoSources are not working, one has to ensure that the necessary libs are located as git repositories in the benchbuild/tmp dir.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

File: varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        version_source = local.path(self.source_of_primary)
+        c_compiler = bb.compiler.cc(self)  # type: ignore
+        cxx_compiler = bb.compiler.cxx(self)  # type: ignore
+        mkdir(version_source / "build")
+
+        # As long as multiple VaraTestRepoSources are not working, one has to
+        # ensure that the necessary libs are located as git repositories in the
+        # benchbuild/tmp dir
+
+        path_to_libs_in_tmp = version_source / "../../../../../tmp"
```
Comment (vulder) [inline varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py]: You can get the path to the BB tmp dir from the BB config

**Review by vulder — APPROVED** [review]
lgtm

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/varats/projects/cpp_projects/TwoLibsOneProjectInteraction_DiscreteLibs_SingleProject.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5977264&bid=19599711)

**Pull request merged**